x/costaking/keeper/hooks_staking.go

```diff
@@ -0,0 +1,127 @@
+//
+// State Changes:
+// - ActiveBaby += (new_amount - old_amount)
+// - If differece is negative, ActiveBaby is subtracted
+//
+// Note: This hook uses a cache to track previous delegation amounts to calculate the delta.
+func (h HookStaking) AfterDelegationModified(ctx context.Context, delAddr sdk.AccAddress, valAddr sdk.ValAddress) error {
+	del, err := h.k.stkK.GetDelegation(ctx, delAddr, valAddr)
+	if err != nil { // we stop if the delegation is not found, because it must be found
+		return err
+	}
+
+	delTokens, err := h.k.TokensFromShares(ctx, valAddr, del.Shares)
+	if err != nil {
+		return err
+	}
+
+	delTokensBefore := h.k.stkCache.GetStakedAmount(delAddr, valAddr)
+	delTokenChange := delTokens.Sub(delTokensBefore).TruncateInt()
+	return h.k.costakerModified(ctx, delAddr, func(rwdTracker *types.CostakerRewardsTracker) {
+		rwdTracker.ActiveBaby = rwdTracker.ActiveBaby.Add(delTokenChange)
+	})
+}
+
+// BeforeDelegationSharesModified handles pre-modification state caching.
+// This hook is triggered before a cosmos staking delegation is modified. It caches
+// the current delegation amount so that AfterDelegationModified can calculate the delta.
+//
+// State Changes:
+// - Caches current delegation amount in temporary storage
+func (h HookStaking) BeforeDelegationSharesModified(ctx context.Context, delAddr sdk.AccAddress, valAddr sdk.ValAddress) error {
+	del, err := h.k.stkK.GetDelegation(ctx, delAddr, valAddr)
+	if err != nil {
+		// probably is not found, but we don't want to stop execution for this
+		h.k.Logger(ctx).Error("hook costaking BeforeDelegationSharesModified", err)
+		return nil
+	}
+
+	delTokens, err := h.k.TokensFromShares(ctx, valAddr, del.Shares)
+	if err != nil {
+		return err
+	}
+	h.k.stkCache.SetStakedAmount(delAddr, valAddr, delTokens)
+	return nil
+}
+
+// AfterUnbondingInitiated implements types.StakingHooks.
+func (h HookStaking) AfterUnbondingInitiated(ctx context.Context, id uint64) error {
+	return nil
+}
+
+// AfterValidatorBeginUnbonding implements types.StakingHooks.
+func (h HookStaking) AfterValidatorBeginUnbonding(ctx context.Context, consAddr sdk.ConsAddress, valAddr sdk.ValAddress) error {
+	return nil
+}
+
+// AfterValidatorBonded implements types.StakingHooks.
+func (h HookStaking) AfterValidatorBonded(ctx context.Context, consAddr sdk.ConsAddress, valAddr sdk.ValAddress) error {
+	return nil
+}
+
+// AfterValidatorCreated implements types.StakingHooks.
+func (h HookStaking) AfterValidatorCreated(ctx context.Context, valAddr sdk.ValAddress) error {
+	return nil
+}
+
+// AfterValidatorRemoved implements types.StakingHooks.
+func (h HookStaking) AfterValidatorRemoved(ctx context.Context, consAddr sdk.ConsAddress, valAddr sdk.ValAddress) error {
+	return nil
+}
+
+// BeforeDelegationCreated implements types.StakingHooks.
+func (h HookStaking) BeforeDelegationCreated(ctx context.Context, delAddr sdk.AccAddress, valAddr sdk.ValAddress) error {
+	return nil
```

Comment:
hmm shouldn't we also implement this hooks ? It seems that when this is first delegation from delegator to validator this hook is used - https://github.com/cosmos/cosmos-sdk/blob/v0.50.14/x/staking/keeper/delegation.go#L892, instead of `BeforeDelegationSharesModified`

Reply:
if it is the first delegation we don't need to implement BeforeDelegationCreated, as AfterDelegationModified handles the case with the cache value returning zero if not found